File: java/leetcode/stackQueue/valid_parentheses.java
/**
 * 20. Valid Parentheses
 * Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Note that an empty string is also considered valid.
 */

public class valid_parentheses {

    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                char cStack = stack.pop();
                boolean b1 = c == ')' && cStack != '(';
                boolean b2 = c == '}' && cStack != '{';
                boolean b3 = c == ']' && cStack != '[';
                if (b1 || b2 || b3) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
}

File: golang/leetcode/stackqueue/valid_parentheses.go
package stackqueue

// 20. Valid Parentheses

func isValid(s string) bool {
	stack := make([]rune, 0)
	for _, c := range s {
		if c == '(' || c == '{' || c == '[' {
			stack = append(stack, c)
		} else {
			if len(stack) == 0 {
				return false
			}
			p := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			b1 := c == ')' && p != '('
			b2 := c == '}' && p != '{'
			b3 := c == ']' && p != '['
			if b1 || b2 || b3 {
				return false
			}
		}
	}
	return len(stack) == 0
}
